### PR TITLE
Added some exampled of heatmaps

### PR DESCRIPTION
**Pull request by TuomasBorman**
Hi,

here are some examples of heatmaps.

-Tuomas

### CODE REVIEW

File: 16_microbiome_community.Rmd [hunk 1]
```diff
@@ -49,24 +49,220 @@ Community composition can be visualized with heatmap where other axis represent
...
+Z transformation is applied to CLR-transformed data. After that abundances are 
+plotted at Phylum level. 
 
 ```{r heatmap}
 library(ggplot2)
+# Does clr-transformation
+tse_phylum <- transformSamples(tse_phylum, method = "clr", pseudocount = 1)
 # Does z-transformation
-tse_phylum <- transformFeatures(tse_phylum, method = "z", name = "Ztransform")
+tse_phylum <- transformFeatures(tse_phylum, method = "z", name = "clr_z")
```
Comment (ChouaibB) [inline 16_microbiome_community.Rmd:61]: Should in this case  'abund_values="clr" ' be mentioned?
When I checked the documentation it uses "counts" as default.

File: 16_microbiome_community.Rmd [hunk 2]
```diff
@@ -49,24 +49,220 @@ Community composition can be visualized with heatmap where other axis represent
...
+}
+library(pheatmap)
+
+# Takes subset: only samples from feces, skin, or tongue
+tse_phylum_subset <- tse_phylum[ , colData(tse_phylum)$SampleType %in% c("Feces", "Skin", "Tongue") ]
+
+# Does clr-transformation
+tse_phylum_subset <- transformSamples(tse_phylum_subset, method = "clr", pseudocount = 1)
+# Does z-transformation
+tse_phylum_subset <- transformFeatures(tse_phylum_subset, method = "z", name = "clr_z")
```
Comment (ChouaibB) [inline 16_microbiome_community.Rmd:96]: similar as earlier, shall the "clr" assay be mentioned as the abundance in the transformation?

**Comment by ChouaibB** [comment]
Really nice visualizations about heatmaps and clustering. The cross-correlation section is a very good example in my opinion, which adds up also an example of using multiassayexperiment in the book.

File: 16_microbiome_community.Rmd (hunk 1 above)
Comment (TuomasBorman) [reply 16_microbiome_community.Rmd:61]: Yes, good. Thanks!

Comment (TuomasBorman) [reply 16_microbiome_community.Rmd]: Fixed

File: 16_microbiome_community.Rmd (hunk 2 above)
Comment (TuomasBorman) [reply 16_microbiome_community.Rmd]: Fixed

**Review by ChouaibB — APPROVED** [review]

File: 16_microbiome_community.Rmd [hunk 3]
```diff
@@ -49,24 +49,222 @@ Community composition can be visualized with heatmap where other axis represent
...
+    install.packages("reshape2")
+}
+# Gets taxa that has at least one statistically significant correlation
+taxa <- as.character(
+  unique(correlation_table[correlation_table[["p.adj"]] < 0.05 &
+                             abs(correlation_table[["Correlation"]]) > 0, "X1"]))
+# Gets taxa that has at least one statistically significant correlation
+lipids <- as.character(
+  unique(correlation_table[correlation_table[["p.adj"]] < 0.05 &
+                             abs(correlation_table[["Correlation"]]) > 0, "X2"]))
```
Comment (antagomir) [inline 16_microbiome_community.Rmd:229]: Why not this:

`correlation_table %>% filter(p.adj < 0.05 & abs(Correlation) > 0) %>% select(X2) %>% unique %>% as.character`

File: 16_microbiome_community.Rmd [hunk 4]
```diff
@@ -49,24 +49,222 @@ Community composition can be visualized with heatmap where other axis represent
...
+
+# Takes only those taxa and lipids that has statistically significant values
+correlation_table <- correlation_table[correlation_table[["X1"]] %in% taxa & 
+                                           correlation_table[["X2"]] %in% lipids, ]
```
Comment (antagomir) [inline 16_microbiome_community.Rmd:233]: see abive

File: 16_microbiome_community.Rmd [hunk 5]
```diff
@@ -49,24 +49,222 @@ Community composition can be visualized with heatmap where other axis represent
...
+Manipulates and reorders the table
+
+```{r cross-correlation4}
+if(!require(reshape2)){
+    install.packages("reshape2")
+}
+# Gets taxa that has at least one statistically significant correlation
+taxa <- as.character(
+  unique(correlation_table[correlation_table[["p.adj"]] < 0.05 &
+                             abs(correlation_table[["Correlation"]]) > 0, "X1"]))
```
Comment (antagomir) [inline 16_microbiome_community.Rmd:225]: see below for alternative filtering command with pipes

File: 16_microbiome_community.Rmd [hunk 6]
```diff
@@ -49,24 +49,222 @@ Community composition can be visualized with heatmap where other axis represent
...
+
+# Creates a ggplot object
+ggplot(correlation_table, aes(x = X1, y = X2, fill = Correlation)) +
+  geom_tile() +
+  scale_fill_gradientn(name = "Correlation", 
+                       breaks = breaks, limits = limits, colours = colours) + 
+  theme(text = element_text(size=10),
+        axis.text.x = element_text(angle=45, hjust=1),
+        legend.key.size = unit(1, "cm")) +
+  labs(x = "Taxa", y = "Lipids")
```
Comment (antagomir) [inline 16_microbiome_community.Rmd:265]: Can we include a line that shows how to add star on the significant associations?

File: 16_microbiome_community.Rmd [hunk 7]
```diff
@@ -49,24 +49,222 @@ Community composition can be visualized with heatmap where other axis represent
...
+order_lipids <- rownames(mat)[lipids_indices]
+
+# Converts taxa and lipids columns to factor so that they have desired order
+correlation_table[["X1"]] <- factor(correlation_table[["X1"]], levels = order_taxa)
+correlation_table[["X2"]] <- factor(correlation_table[["X2"]], levels = order_lipids)
+```
+
+Creates the heatmap
+
+```{r cross-correlation5}
```
Comment (antagomir) [inline 16_microbiome_community.Rmd:251]: Tune fig height so that y axis labels do not overlap so it looks good layout

**Review by antagomir — CHANGES_REQUESTED** [review]
Overall, very nice, thanks!

The cross-correlation example is relatively long, compared to the simple nature of the operation (cross-correlate -> arrange -> visualize). We often need two-assay cross-correlation calculation and plotting for exploratory purposes. Should we have a wrapper function that would allow the user to plot a cross-correlation heatmap between two selected assays (can be from altExp or MAE as well), with relevant arguments on significance/effectsize based filtering etc. that we have here in this example? This would allow the user to plot easily cross-correlations between different data tables within a given TreeSE object. When this is not sufficient, the user can always do more complex analyses by separating the components.

File: 16_microbiome_community.Rmd (hunk 3 above)
Comment (TuomasBorman) [reply 16_microbiome_community.Rmd:229]: This is easier to understand, changed

File: 16_microbiome_community.Rmd (hunk 5 above)
Comment (TuomasBorman) [reply 16_microbiome_community.Rmd:225]: This is easier to understand, changed

File: 16_microbiome_community.Rmd (hunk 4 above)
Comment (TuomasBorman) [reply 16_microbiome_community.Rmd:233]: ok

File: 16_microbiome_community.Rmd (hunk 6 above)
Comment (TuomasBorman) [reply 16_microbiome_community.Rmd:265]: Added

File: 16_microbiome_community.Rmd (hunk 7 above)
Comment (TuomasBorman) [reply 16_microbiome_community.Rmd:251]: height adjusted

**Comment by TuomasBorman** [comment]
> Overall, very nice, thanks!
> 
> The cross-correlation example is relatively long, compared to the simple nature of the operation (cross-correlate -> arrange -> visualize). We often need two-assay cross-correlation calculation and plotting for exploratory purposes. Should we have a wrapper function that would allow the user to plot a cross-correlation heatmap between two selected assays (can be from altExp or MAE as well), with relevant arguments on significance/effectsize based filtering etc. that we have here in this example? This would allow the user to plot easily cross-correlations between different data tables within a given TreeSE object. When this is not sufficient, the user can always do more complex analyses by separating the components.

Hmm, without further thinking, that could be a nice functionality. Should we also implement `microbiome::associate()` to mia?

**Comment by antagomir** [comment]
The associate function would go well together with visual explorations. 

The potential problem is that there are many different use cases with different sample/feature numbers, types of assays etc. and it may be hard to come up with a generic solution. But if we keep it simple with nonparametric Spearman correlations to start with, and user has all freedom to specify assays and transformations with other methods, then I think it should do, this is common routine in exploratory phase. 

This was already discussed in [mia issue #65](https://github.com/microbiome/mia/issues/65), however. We should revisit that and check the capabilities in the cited `MOFA2` to see if that can be readily used. I think Spearman correlations for associations would work for exploratory purposes regardless of assay and size considerations.

**Comment by antagomir** [comment]
I merge this now, so we can have the current examples at least. We can continue the discussion in issue #65 and others. If you have complementary material that is not yet added, you can then make a new PR.

**Pull request merged**